libs/core/algorithms/include/hpx/parallel/util/detail/chunk_size_iterator.hpp

```diff
@@ -71,20 +75,58 @@ namespace hpx { namespace parallel { namespace util { namespace detail {
                 chunk_size_ == other.chunk_size_;
         }
 
-        HPX_HOST_DEVICE typename base_type::reference dereference() const
+        HPX_HOST_DEVICE typename base_type::reference dereference()
+            const noexcept
         {
             return data_;
         }
 
         HPX_HOST_DEVICE void increment()
         {
             // prepare next value
+            HPX_ASSERT(count_ >= chunk());
+
             count_ -= chunk();
 
             iterator() = parallel::v1::detail::next(iterator(), chunk());
             chunk() = (hpx::detail::min)(chunk_size_, count_);
         }
 
+        HPX_HOST_DEVICE void decrement()
+        {
+            // prepare next value
+            count_ += chunk();
```

Comment:
I might not be right. I'm aware of this potential issue, but since all we do with this iterator is to possibly `std::advance` it, I decided to ignore it. I'll take a note and will try to work on it next.

Reply:
If we only `advance` it do we need `decrement` for something?

Reply:
It's there only for completeness's sake.

Reply:
Can we check for the bad case somehow? This will be a hard-to-debug issue for someone in the future way after we've forgotten about this behaviour and it'd be nice to fail early here if it's ever used.

Reply:
Ok, I'll try to make it right with this PR. I'll also add a test.

Reply:
The iterators should be fine now.